### PR TITLE
fix: allow to override partially the setup config (#10229)

### LINKED ISSUE #10229: man dir does not change as specified by --mandir
## Problem

dune has been failing to build from source on **pkgsrc** because they can't get dune to put the man pages in a different location (`share/man/*`) from dune's default `man/*` (https://us-central.manta.mnx.io/pkgsrc/public/reports/upstream-trunk/20240305.1526/ocaml-dune-3.11.1/install.log).

As an experiment, I cloned dune from github to build it the make way. I executed the following sequence, from what I understood through various conversations on the github repo:

```
cd dune/
./configure --mandir=./fake_root/share/man
PREFIX=fake_root make release
PREFIX=fake_root make install
```

`configure` command seem to be doing as it is intended, and it creates/modifies `src/dune_rules/setup.ml`:

```
let library_path = []

let roots : string option Install.Roots.t =
  { lib_root = None
  ; man = Some "/Users/jb/projects/dune/./fake_root/share/man"
  ; doc_root = None
  ; etc_root = None
  ; share_root = None
  ; bin = None
  ; sbin = None
  ; libexec_root = None
  }
```

## Expected Behavior

man pages should be installed at `fake_root/share/man/*`.

## Actual Behavior

man pages are still installed at man/* instead of share/man/*.

```
> tree -d fake_root
fake_root
├── bin
├── doc
│   └── dune
│       └── odoc-pages
├── lib
│   └── dune
├── man
│   ├── man1
│   └── man5
└── share
    └── emacs
        └── site-lisp
```

## Reproduction

1. Clone dune repo
2. Execute the following sequence of commands.

```
cd dune/
./configure --mandir=./fake_root/share/man
PREFIX=fake_root make release
PREFIX=fake_root make install
```

## Specifications

- Version of `dune` (output of `dune --version`): `main` branch
- Version of `ocaml` (output of `ocamlc --version`): 4.14.1
- Operating system (distribution and version): macOS Sonoma 14.3 (Silicon aarm64 architecture)

### PR DESCRIPTION
**Pull request by moyodiallo**
Fixes #10229

### CODE REVIEW

**Comment by moyodiallo** [comment]
Done in https://github.com/ocaml/dune/commit/f39754471b057b26cc4c211a472142257be77533

File: bin/install_uninstall.ml [hunk 1]
```diff
@@ -35,12 +35,14 @@ let get_dirs context ~prefix_from_command_line ~from_command_line =
...
   let module Roots = Install.Roots in
   let prefix_from_command_line = Option.map ~f:Path.of_string prefix_from_command_line in
+  let setup_roots = Roots.map ~f:(Option.map ~f:Path.of_string) Dune_rules.Setup.roots in
   let+ roots =
     match prefix_from_command_line with
     | None -> Memo.run (Context.roots context)
     | Some prefix ->
       Roots.opam_from_prefix prefix ~relative:Path.relative
       |> Roots.map ~f:(fun s -> Some s)
+      |> Roots.first_has_priority setup_roots
```
Comment (rgrinberg) [inline bin/install_uninstall.ml:45]: Why do you need this line?

**Comment by rgrinberg** [comment]
Can you add a test case reproducing bug in a separate PR?

**Comment by moyodiallo** [comment]
> Can you add a test case reproducing bug in a separate PR?

Do we have those kind of tests ? The setup_roots are hard-coded in `src/dune_rules/setup.ml`.

**Comment by moyodiallo** [comment]
> build commands are taken from `dune.opam`; I'm not sure which ones `pkgsrc` uses)

I was missing this detail in the context of the PR. It's fixable. I'll add a test reproducing the bug in a separate PR and rework the PR. Thanks @emillon.

**Comment by emillon** [comment]
I don't know what's the right solution TBH. the original command uses a relative path for mandir, does that mean it should be interpreted relative to prefix? it would be fine to say that we don't accept relative paths here, for example. before jumping to a new implementation and look for a fix at all costs, I think we should answer these questions in order:
- is the original request something we want to support?
- how do we want to support that?
- how do implement that?

**Comment by moyodiallo** [comment]
> I don't know what's the right solution TBH. the original command uses a relative path for mandir, does that mean it should be interpreted relative to prefix? it would be fine to say that we don't accept relative paths here, for example. before jumping to a new implementation and look for a fix at all costs, I think we should answer these questions in order:
>
> * is the original request something we want to support?
> * how do we want to support that?
> * how do implement that?

The exact command is (from this [link](https://us-central.manta.mnx.io/pkgsrc/public/reports/upstream-trunk/20240305.1526/ocaml-dune-3.11.1/install.log) in the issue) : 
```
./_boot/dune.exe install --prefix /opt/pkg  --destdir /home/pbulk/build/devel/ocaml-dune/work/.destdir dune
```
I think if we give the ability to have relative dir for other arguments(--mandir,--docdir,...) when `--destdir` is mentioned. This could solve the issue. 

I've tried this:
```
 |  $ echo "(lang dune 2.0)" > dune-project
 |  $ touch foo.opam manfile
 |  $ cat >dune <<EOF
 |  > (install
 |  >  (section man)
 |  >  (files manfile))
 |  > EOF
 |  $ dune build @install
 |  $ mkdir dest-dir
 |  $ dune install --prefix foo --destdir $PWD/dest-dir --mandir $PWD/dest-dir/foo/man/man1
  |  $ tree . -a
+|  .
+|  |-- _build
+|  |   |-- .db
+|  |   |-- .digest-db
+|  |   |-- .filesystem-clock
+|  |   |-- .lock
+|  |   |-- default
+|  |   |   |-- .dune
+|  |   |   |   |-- configurator
+|  |   |   |   `-- configurator.v2
+|  |   |   |-- META.foo
+|  |   |   |-- foo.dune-package
+|  |   |   |-- foo.install
+|  |   |   |-- foo.opam
+|  |   |   `-- manfile
+|  |   |-- install
+|  |   |   `-- default
+|  |   |       |-- lib
+|  |   |       |   `-- foo
+|  |   |       |       |-- META -> ../../../../default/META.foo
+|  |   |       |       |-- dune-package -> ../../../../default/foo.dune-package
+|  |   |       |       `-- opam -> ../../../../default/foo.opam
+|  |   |       `-- man
+|  |   |           `-- manfile -> ../../../default/manfile
+|  |   `-- log
+|  |-- dest-dir
+|  |   |-- foo
+|  |   |   `-- lib
+|  |   |       `-- foo
+|  |   |           |-- META
+|  |   |           |-- dune-package
+|  |   |           `-- opam
+|  |   `-- home
+|  |       `-- alpha
+|  |           `-- hack
+|  |               `-- moyodiallo
+|  |                   `-- dune
+|  |                       `-- _build
+|  |                           `-- .sandbox
+|  |                               `-- 56db160efae28095c6a5b776b852d9c6
+|  |                                   `-- default
+|  |                                       `-- test
+|  |                                           `-- blackbox-tests
+|  |                                               `-- test-cases
+|  |                                                   `-- install-dir
+|  |                                                       `-- dest-dir
+|  |                                                           `-- foo
+|  |                                                               `-- man
+|  |                                                                   `-- man1
+|  |                                                                       `-- manfile
+|  |-- dune
+|  |-- dune-project
+|  |-- foo.opam
+|  `-- manfile
+|
```

`home/../../ ../_build/.../.. ../man/man1/manfile` directory appears in `dest-dir` what is following the logic because the destination is known, but with a bug. In the other hand `--mandir` with relative dir gives : 
```
dune: option '--mandir': the path must be absolute to avoid ambiguity
+|  Usage: dune install [OPTION]… [PACKAGE]…
+|  Try 'dune install --help' or 'dune --help' for more information.
```

**Comment by rgrinberg** [comment]
I remember this being discussed in the team meeting. The verdict was: all paths passed in the command line should be interpreted to relative to CWD

**Pull request closed without merging**